---
title: "Lab8: Sequence Alignment, Phylogenetic Analysis and Tree Visualization"
author: "Manya Vivek"
date: "10/19/2020"
---

# Exercise 1 

* Open NGPhylogeny.fr. 

* Click on "Let's go! with One Click Workflow" 

* Choose FASTME/OneClick in Select your workflow 

* Either input the "tree phylogenetic tree" or copy and paste the amino acid sequence from the file downloaded. 

* Hit submit 

* Download output tree file 

![TreeImage](Tree_image.png) 

```{r}
library(tidyverse)
library(ggtree)
library(treeio)
library(ggimage)
```

# Exercise 2 

[TOL_output.mafft](data/TOL_output.mafft)

```{r}
tt <- read.tree("data/fastree_result.tre")
tt
```
```{r, fig.height = 15, fig.width = 20}
ggtree(tt) + geom_tiplab()
```



# Exercise 3
```{r}
library(tidyverse)
library(ggtree)
library(treeio)
library(ggimage)
```

```{r}
tree <- read.tree("data/tree_newick.nwk")
tree
```

```{r}
ggplot(tree) + geom_tree() + theme_tree()
```

```{r}
ggtree(tree)
```

# adding a scale
```{r}
ggtree(tree) + geom_treescale()
```
```{r}
ggtree(tree) + theme_tree2()
```
```{r}
ggtree(tree, branch.length="none")
```
```{r}
ggtree(tree, branch.length="none", color="blue", size=2, linetype=3)
```


##### Exercise 1 
```{r}
ggtree(tree, layout = "slanted")
```

```{r}
ggtree(tree, layout = "circular")
```

```{r}
ggtree(tree, branch.length="none", color="red", size=2, linetype=1)
```

#-----------------------------------------------------------------------------------------------

### Other tree genomes 

```{r}
# create the basic plot
p <- ggtree(tree)
```

##### Node points, tip points and tiplab 
```{r}
p + geom_nodepoint() + geom_tippoint() + geom_tiplab()
```

##### Exercise 2 

```{r}
p + geom_tippoint(shape = "diamond", color = "purple") + geom_tiplab(color = "purple") + geom_nodepoint(color = "yellow", alpha = 0.5, size = 3)
```

# Tree annotation

```{r}
ggtree(tree) + geom_text(aes(label=node), hjust=-.1)
```
```{r}
ggtree(tree) + geom_tiplab()
```

# Labelling Clades


```{r}
ggtree(tree) + 
  geom_cladelabel(node=17, label="Random", color="red", offset =.8, align = TRUE)+
  geom_cladelabel(node=21, label="Different", 
                  color="blue", offset=.8, align = TRUE) +
 theme_tree2() + 
  xlim(0, 70) 
```

### removing x-axis scale 
```{r}
ggtree(tree) + 
  geom_cladelabel(node=17, label="Random", color="red", offset =.8, align = TRUE)+
  geom_cladelabel(node=21, label="Different", 
                  color="blue", offset=.8, align = TRUE) +
 theme_tree2() + 
  xlim(0, 70) + 
  theme_tree()
```

###  highlight the entire clade 
```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_hilight(node=17, fill="gold") + 
  geom_hilight(node=21, fill="purple")
```

```{r}
ggtree(tree) + geom_tiplab()
MRCA(tree, c("C", "E"))
MRCA(tree, c("G", "H"))
```

# Connecting Taxa 
#### (use geom_taxalink())
```{r}
ggtree(tree) + 
  geom_tiplab() + 
  geom_taxalink("E", "H", color="blue3") +
  geom_taxalink("C", "G", color="orange2", curvature=-.9)
```

##### Exercise 3

```{r}
ggtree(tree) + geom_text(aes(label=node), hjust=-.3)
```



```{r}
ggtree(tree) 
MRCA(tree, c("B", "C"))
MRCA(tree, c("L", "J"))
```

```{r}
ggtree(tree) + 
  geom_tiplab() +
  geom_hilight(node=19, fill="sky blue") + 
  geom_hilight(node=23, fill="lavender") + 
  geom_taxalink("C", "E", linetype = 2) +
  geom_taxalink("G", "J", linetype = 2) + 
  theme_tree2() + 
  ggtitle("Practice Phylogenetic tree")
```

```{r}
  ggtree(tree, layout = "circular") +
  geom_tiplab() +
  geom_hilight(node=19, fill="sky blue") + 
  geom_hilight(node=23, fill="lavender") + 
  theme_tree2() + 
  ggtitle("Practice Phylogenetic circular tree") 
```


### Read the data
```{r}
tree <- read.beast("data/flu_tree_beast.tree")
tree
```
### Scale Bar
```{r}
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() 
```

### Finally, add tip labels and adjust axis
```{r, fig.height = 15, fig.width = 10}
ggtree(tree, mrsd="2013-01-01") + 
  theme_tree2() + 
  geom_tiplab(align=TRUE, linesize=.5) + 
  xlim(1990, 2020)
```

```{r, fig.height = 15, fig.width = 10}
msaplot(p=ggtree(tree), fasta="data/flu_aasequence.fasta", window=c(150, 175))
```
```{r}
set.seed(42)
trees <- lapply(rep(c(10, 25, 50, 100), 3), rtree)
class(trees) <- "multiPhylo"
ggtree(trees) + facet_wrap(~.id, scale="free", ncol=4) + ggtitle("Phylogenetic trees")
```

```{r}
tree <- rtree(30)
p <- ggtree(tree)
d1 <- data.frame(id=tree$tip.label, val=rnorm(30, sd=3))
p2 <- facet_plot(p, panel="dot", data=d1, geom=geom_point, aes(x=val), color='red3')
d2 <- data.frame(id=tree$tip.label, value = abs(rnorm(30, mean=100, sd=50)))
p3 <- facet_plot(p2, panel='bar', data=d2, geom=geom_segment, 
           aes(x=0, xend=value, y=y, yend=y), size=3, color='blue4') 
p3 + theme_tree2()
```


----------------------------------------------------------------------------------------------------------------------------

[Tree File](data/fastree_result.tre)

```{r}
tree <- read.tree("data/fastree_result.tre")
tree
```
```{r}
ggtree(tree) + 
  theme_tree2() +
    geom_tiplab() + 
  geom_nodelab() + 
  xlim(0,2) 
```


```{r}
ggtree(tree, layout = "circular") + 
  theme_tree2() +
    geom_tiplab() + 
  geom_nodelab() + 
  xlim(0,36)
```

# Exercise 4 

```{r}
library(ggtree)
```
[Output_tree.nhx](data/Output_Tree.nhx)


```{r}
nhxfile <- read.tree("data/Output_Tree.nhx")
nhxfile
```

```{r}
ggtree(nhxfile) + geom_text(aes(label=node), hjust=-.3)
```

```{r}
ggtree(nhxfile) +
    geom_nodelab()+
  geom_tiplab(hjust=-.1, size = 3) +
    geom_hilight(node=23, fill="gold", extend = 0.5) + 
    geom_hilight(node=37, fill="purple", extend = 0.8) + 
    geom_hilight(node=27, fill="blue", extend = 0.5) 
xlim(0,36)
```

```{r}
ggtree(nhxfile, layout = "circular") +
    geom_nodelab()+
  geom_tiplab(hjust=-.1, size = 3) +
    geom_hilight(node=23, fill="gold", extend = 0.5) + 
    geom_hilight(node=37, fill="purple", extend = 0.5) + 
    geom_hilight(node=27, fill="blue", extend = 0.5) 
xlim(0,36)
```


